### PR TITLE
prowjob attempt

### PR DESCRIPTION
**Pull request by MHBauer**
I have almost no idea what I am doing.

Is there a way to experiment with this without pushing commits? Seems heavyweight.

/cc @cjwagner @BenTheElder

### CODE REVIEW

**Comment by MHBauer** [comment]
```
	job_config_test.go:188: kubernetes-incubator/service-catalog in svc-cat-echo rerun_command: /test echo != expected: /test svc-cat-echo
```
What's the point of specifying this if it has to be an exact match for the name?

**Comment by BenTheElder** [comment]
@stevekuznetsov speaking of that, forgot to bring this up as tech debt at the meeting the other day, WDYT? I think we should probably drop needing to specify a `/test` regex

**Comment by BenTheElder** [comment]
It seems github ate my email reply, copying here:
> That's actually up for debate, Prow itself doesn't enforce this, but test-infra has. I think the plan is to not require specifying it soon. This is one of the tech debt bits on @cjwagner's radar.

File: prow/config.yaml [hunk 1]
```diff
@@ -6455,6 +6455,27 @@ presubmits:
...
+    agent: kubernetes
+    context: svc-cat-echo
+    always_run: true
+    rerun_command: "/test svc-cat-echo"
+    trigger: "(?m)^/test (all|svc-cat-echo)\\s*"
+    decorate: true
+    spec:
+      containers:
+      - image: alpine
+        command:
```
Comment (BenTheElder) [inline prow/config.yaml]: pretty sure this should be:
```yaml
command:
- sh
args:
- -c
- env; ls; echo "The artifacts dir is ${ARTIFACTS}"
```

**Comment by BenTheElder** [comment]
> Is there a way to experiment with this without pushing commits? Seems heavyweight.

If you're running a prow yes,`prow/cmd/mkpj`, but I don't think we should have people running arbitrary tasks on the kubernetes CI without any review, so we just PR the definitions..?

**Comment by fejta** [comment]
> What's the point of specifying this if it has to be an exact match for the name?

we validate that `/test exact-job-name` triggers the job, because this is easy for people who didn't write the job to understand (I can trigger context blah-blah with a `/test blah-blah` command).

The regex can also match in other ways, if that is necessary for some reason.

**Comment by fejta** [comment]
> Is there a way to experiment with this without pushing commits?

`docker run`

File: prow/config.yaml (hunk 1 above)
Comment (MHBauer) [reply prow/config.yaml]: okay, not sure what the rules are, trying to figure it out.
this one was riffing off of the job
```
tensorflow/minigo: 
 - name: pull-tf-minigo-cc
```
which has 
```
        command:
        - /bin/bash
        - -c
```
but I do also see now that it has the args as `&&` together of commands.

**Comment by MHBauer** [comment]
I winged it and ended up with this output from `mkpj`, I'll throw it into a kubernetes and see what happens.

Is there an example of this documented somewhere? Probably on a test-infra call in the past?

```console
$ ./mkpj -job svc-cat-echo -config-path ./prow/config.yaml 
Base ref (e.g. master): 
Base SHA (e.g. 72bcb5d80): 
PR Number: 
PR author: 
PR SHA (e.g. 72bcb5d80): 
apiVersion: prow.k8s.io/v1
kind: ProwJob
metadata:
  creationTimestamp: null
  name: 1128bbcb-78e2-11e8-9249-54e1ad543ebd
spec:
  agent: kubernetes
  cluster: default
  context: svc-cat-echo
  decoration_config:
    gcs_configuration:
      bucket: kubernetes-jenkins
      default_org: kubernetes
      default_repo: kubernetes
      path_strategy: legacy
    gcs_credentials_secret: service-account
    grace_period: 15000000000
    timeout: 7200000000000
    utility_images:
      clonerefs: registry.svc.ci.openshift.org/ci/clonerefs:latest
      entrypoint: registry.svc.ci.openshift.org/ci/entrypoint:latest
      initupload: registry.svc.ci.openshift.org/ci/initupload:latest
      sidecar: registry.svc.ci.openshift.org/ci/sidecar:latest
  job: svc-cat-echo
  pod_spec:
    containers:
    - args:
      - env
      - ;
      - ls
      - ;
      - The artifacts dir is $(ARTIFACTS)
      command:
      - sh
      - -c
      image: alpine
      name: ""
      resources: {}
  refs:
    org: kubernetes-incubator
    pulls:
    - {}
    repo: service-catalog
  report: true
  rerun_command: /test svc-cat-echo
  type: presubmit
status:
  startTime: 2018-06-26T01:41:38Z
  state: triggered
```

**Comment by BenTheElder** [comment]
mkpj is something of a power user tool for Prow at the moment. ProwJob CRDs
are normally generated by one of the other components. That said there's
not much to document, it consumes the config and job name and spits out a
CRD to stdout.

**Comment by MHBauer** [comment]
I see now that it's a CRD, so I'll try and set up prow in my cluster.

File: prow/config.yaml (hunk 1 above)
Comment (MHBauer) [reply prow/config.yaml]: note to self: go try and read the prow code to see what the prow spec is transformed into, i.e. a podspec or containerspec, etc.

**Comment by MHBauer** [comment]
Not sure I'll be able to, does the prow deployment require gce?

File: prow/config.yaml (hunk 1 above)
Comment (BenTheElder) [reply prow/config.yaml]: It's a PodSpec

**Comment by BenTheElder** [comment]
> Not sure I'll be able to, does the prow deployment require gce?

No, Prow runs on any k8s cluster, though the cold-storage (artifact upload) / logging tools use GCS to do permanent storage. OpenShift runs one on an OpenShift cluster somewhere.

**Comment by cjwagner** [comment]
You don't need to set up your own Prow instance just to test running a job.
Typically we try out new jobs by setting `skip_report: true` so that the job doesn't report a status context to GitHub and instead just uploads results to GCS. If the job is broken or misconfigured it won't affect anyone.

File: prow/config.yaml (hunk 1 above)
Comment (MHBauer) [reply prow/config.yaml]: yes, exactly so:
https://github.com/kubernetes/test-infra/blob/1bb0838c9b49beb86d4c25a795a0a2f3ec317f0c/prow/kube/prowjob.go#L145-L147

Comment (MHBauer) [reply prow/config.yaml]: who is specifying the containers `Name` field?

**Comment by MHBauer** [comment]
the bootstrap images seems to be the one I want. going to make the changes and push for review, as I'm not sure how to run it. I seems to fall all the way back to `bootstrap.py` in the end, which it sounded like something I was supposed to avoid.

looking through the dockerfile I'm wondering if docker can be updated since the issue referred to is closed.
https://github.com/kubernetes/test-infra/blob/master/images/bootstrap/Dockerfile#L99-L100

in some timeline I'm a docker engine maintainer, so maybe something I can help fix or push an issue on.

**Comment by BenTheElder** [comment]
I haven't upgraded docker because the current version works fine for
everyone using it so far and upgrades in the past have caused issues. k8s
also moves pretty slowly to upgrade docker as well.

I'd rather not unpin docker as surprise upgrades when rebuilding the image
have definitely broken us. It should be upgraded at some point, but so far
there doesn't seem to be any major rush.

**Comment by MHBauer** [comment]
https://github.com/kubernetes/test-infra/pull/8431#issuecomment-401990111

I see. For podutils, where would the image come from?
I see the intention here as the image is fairly static, and what it does is have the build environment, and does the checkout of the code into a specific environment, before doing a run.

As it is, most of ourprocess was set up to be as convenient as could be while similar in both dev and test, hence the containerized environment. This process would seem to need to change somewhat.

We do have some support for a non-docker build `NO_DOCKER=1 make test`, but it is not officially supported. What would be the appropriate image to use for a golang build & test? Perhaps we should try that.

We still require some way to build images and push them.

**Comment by MHBauer** [comment]
https://github.com/kubernetes/test-infra/pull/8431#issuecomment-401993526
> The image your dockerized build uses should probably work as well.

Ya, but that image is made using docker, as the first step of the build. As it exists, the build only needs two things, make and docker.

**Comment by BenTheElder** [comment]
This Dockerfile?
https://github.com/kubernetes-incubator/service-catalog/blob/2e426633be0a1e89ddd3de68fd54de4c30fae8c0/build/build-image/Dockerfile

If so, it appears to only have $GO_VERSION as an argument, we could pre-publish this image.

**Comment by MHBauer** [comment]
Yes, that one. Pushing it somewhere is an existing work item that we've never gotten around to. 

That's only part of what's dockerized. We've also got a kubectl image used for the integration tests.

I've pushed an update that uses the base golang image, but that wouldn't be able to run all of the integration tests without docker or some other way to call out to kubectl.

**Comment by MHBauer** [comment]
And I did check that that would work. An approximate docker equivalent is
`docker run --rm -v $(pwd):/go/src/github.com/kubernetes-incubator/service-catalog -w /go/src/github.com/kubernetes-incubator/service-catalog -e NO_DOCKER=1 golang:1.10 make`
Replace the volume mount with the checked out svc-cat code.

File: prow/config.yaml [hunk 2]
```diff
@@ -3844,6 +3844,26 @@ presubmits:
         - "--service-account=/etc/service-account/service-account.json"
         - "--upload=gs://kubernetes-jenkins/pr-logs"
 
+  kubernetes-incubator/service-catalog:
+  - name: svc-cat-test
+    agent: kubernetes
+    context: svc-cat-test
+    always_run: true
```
Comment (BenTheElder) [inline prow/config.yaml]: you might want to add `skip_report: true` so this runs on every PR still, but doesn't post a status or any comments. we can check on prow.k8s.io / k8s-gubernator.appspot.com for results while we make sure this behaves properly, then remove the `skip_report`.

File: prow/config.yaml [hunk 3]
```diff
@@ -3844,6 +3844,26 @@ presubmits:
         - "--service-account=/etc/service-account/service-account.json"
         - "--upload=gs://kubernetes-jenkins/pr-logs"
 
+  kubernetes-incubator/service-catalog:
+  - name: svc-cat-test
```
Comment (BenTheElder) [inline prow/config.yaml]: for whatever reasons k8s jobs usually have `pull|ci|periodic-$name` /shrug

**Comment by BenTheElder** [comment]
two suggestions, feel free to either ignore and `/hold cancel` to go ahead, or follow up and I'll re-stamp

File: prow/config.yaml (hunk 3 above)
Comment (MHBauer) [reply prow/config.yaml]: More than happy to rename to match existing convention. What do the notations mean? Which should I choose?

File: prow/config.yaml [hunk 4]
```diff
@@ -3844,6 +3844,27 @@ presubmits:
...
+    context: pull-service-catalog-unit
+    always_run: true
+    skip_report: true
+    rerun_command: "/test pull-service-catalog-unit"
+    trigger: "(?m)^/test (all|pull-service-catalog-unit)\\s*"
+    decorate: true
+    spec:
+      containers:
+      - image: golang
+        imagePullPolicy: Always
```
Comment (BenTheElder) [inline prow/config.yaml]: I don't think we need or want this?

**Comment by MHBauer** [comment]
updated. thanks for the help.

File: prow/config.yaml (hunk 4 above)
Comment (MHBauer) [reply prow/config.yaml]: okay, one sec.

**Comment by BenTheElder** [comment]
Would you mind squashing your commits? Also I don't think we want `imagePullPolicy: Always`, we don't need to re-pull go onto the node before every service catalog PR?

**Comment by BenTheElder** [comment]
Sorry this has been so lengthy! You're one of the (_the_?) first to create a PR test job using the pod-utilities from scratch with no prior experience! 
We'll hopefully get this ironed out better soon for future onboarding 😬

**Comment by MHBauer** [comment]
I see future work on our part to adapt the build to this environment.

I think there should be a very standard build environment provided for everyone to reuse so that there is a known good env to develop and test in.

**Comment by MHBauer** [comment]
ensquishinated with a nicer commit message

**Comment by BenTheElder** [comment]
> I think there should be a very standard build environment provided for everyone to reuse so that there is a known good env to develop and test in.

Definitely, we've been moving towards that with `images/kubekins-e2e` (as a rough approximation) hosting the majority of the build / unit test / e2e jobs. something a bit cleaner would be more broadly useful but already we needed something approximating that to get consistent remote caching (to avoid different versions of the host-installed tools like `gcc`, `rpmbuild`).

I want to do a v2 of this image for the podutils at some point. Right now I think the focus has been more on maturing the podutils. We've still been finding a few [quirks](https://github.com/kubernetes/test-infra/issues/8509) that need fixing.

**Comment by BenTheElder** [comment]
/hold cancel
👀

**Comment by carolynvs** [comment]
Some of our build and deploy steps make use of secrets to upload artifacts to cloud storage (e.g. helm charts, binaries, docker images). So we'll also need to look into how we can put secrets into prow.

**Comment by BenTheElder** [comment]
yep! we just run things in kubernetes pods so we can mount kube secrets,
that said we should discuss _how_ secret these are, ideally they're just
for managing test resources / artifacts which is the easiest case

**Comment by carolynvs** [comment]
They are storage keys for containers that only hold svcat stuff, like our svcat cli binaries or helm charts. I think we also have some quay credentials too, again just for service catalog things.

**Comment by BenTheElder** [comment]
Are any user consumed images / binaries stored with the same keys though?
For kubernetes itself the user facing and CI / test storage is separate
(users _could_ use the CI storage, but that's on them). This is nicer from
a security POV.

**Comment by carolynvs** [comment]
I'm working on that right now and will take your suggestion to split out the CI storage for our docker images. For our helm charts and binaries, it's all user-facing so that's not an issue.

**Comment by BenTheElder** [comment]
> I'm working on that right now and will take your suggestion to split out the CI storage for our docker images.

Awesome, yeah the main concern is that any job running malicious code (say an evil PR in another repo with /ok-to-test) could potentially pivot to stealing the secrets from the build / test cluster it is in and then exfil them for use publishing malicious binaries / images (or even just a job accidentally logging them for some reason...)

We could even run svcat it it's own build / test cluster, but there's sorta the same issue still.

We don't expect this to happen, but if the CI only has credentials for test resources anyhow, worst case the resources are just wasted for a bit until we clean up, versus worst case users get evil builds.

> For our helm charts and binaries, it's all user-facing so that's not an issue.

what does this mean though?

**Comment by carolynvs** [comment]
Sorry that wasn't clear. The way we have our CI setup right now, there are jobs that run during a PR, and a separate job that only runs on master. That master job has access to our deploy secrets so that it can publish our helm charts, binaries and docker images. The PR job doesn't need access to any of that, and I'll make sure that when I put in docker image pushes between the steps in the PR job that it pushes to a quarantined location, so that if PR secrets are abused, they can't do any damage.

**Comment by BenTheElder** [comment]
Awesome! :D 

We've had a general need for this for a while, what we can do is split the "publish from master" type jobs into a seperate build cluster and pin them to that, which we already have support for now 😄

**Pull request merged**